Prelim Task Issue #243

vaditya04 opened the pull request:
@wkerzendorf @orbitfold 

I've basically adapted the launcher.py file and added grid generation to it. I decided to create a new file (grid_tester.py) with all the required implementation.

This program works, but there is an issue with it:

Cannot run a list of config_dicts simultaneously using queue_parameter_set_list, since simpleworker is an instance method. It is also decorated, so it is not possible to simply unpickle the method and send it into the map function. Since 'interactive' is a decorator within IPython, we cannot modify it with a functool.wrap(). Since I couldn't find a way around this, I'm running the config files using queue_parameter_set in a loop for now.
UPDATE: Fixed this
Thanks

wkerzendorf commented:
okay, this code is a little bit more difficult to review as I need to go back and think about the original intentions of my code. There are two sides to this (1) this PR shows me how you can deal with code and (2) make it useable for the project. Number 1 is very good, for Number 2 I need to think longer than 5 minutes about it. So now, do start your application and tell us what you would with the 3 months you have. Be as specific as possible. Let me know if there are any problems

vaditya04 commented:
@wkerzendorf @orbitfold 
I fixed the pickling error such that the map function can now be used for parallelizing. My solution is to use a more robust pickler, dill http://nbviewer.ipython.org/github/adrn/ipython/blob/master/examples/Parallel%20Computing/Using%20Dill.ipynb

It works like a charm, and the best part is that IPython parallel has a hook for dill (use_dill()). So aside from installing it, all I had to do was call use_dill() on the remote 
b2cae769b4eeafc51260ba1838ebf914a2f0030b
(Please look into the changes in launcher.py)

My grid testing  code is here : https://github.com/vaditya04/dalek/blob/b2cae769b4eeafc51260ba1838ebf914a2f0030b/dalek/parallel/grid_tester.py

wkerzendorf commented:
@vaditya04 currently you are writing a completely new python file with a rewrite of `BaseLauncher` and so on. It's great to have a `GridBuilder` but please use the existing infrastructure.

vaditya04 commented:
@wkerzendorf I changed the file to run using existing infrastructure. I've not removed the GridBuilder class because that implementation is only present in test files, and I thought including test files might not be a good idea.
I rewrote BaseLauncher because I wanted to figure out the pickling issue, forgot to switch back to using launcher.py, sorry about that. 
Also, I made a one line change in the launcher file to fix the pickling  error.
Please let me know if you want me to remove the GridBuilder class as well. 
Thanks

https://github.com/vaditya04/dalek/blob/master/dalek/parallel/grid_tester.py

wkerzendorf commented on dalek/parallel/launcher.py:
```diff
@@ -93,6 +93,7 @@ def __init__(self, remote_clients, worker=simple_worker,
         self.prepare_remote_clients(remote_clients, atom_data)
         self.worker = worker
         self.lbv = remote_clients.load_balanced_view()
+        self.remote_clients[:].use_dill()   #This fixes the pickling error, now map function can be used
```
That looks interesting. good work
